sphinx/source/docs/user_guide/data.rst

```diff
@@ -101,20 +127,49 @@ the ``source`` argument:
     p = figure()
     p.circle(x='x_values', y='y_values', source=source)
 
-To use a ColumnDataSource with a plotting function, you need to pass at least
-these three arguments:
+Modifying a ColumnDataSource
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ``x``: the name of the ColumnDataSource's column that contains the data for
-  the x values of your plot
-* ``y``: the name of the ColumnDataSource's column that contains the data for
-  the y values of your plot
-* ``source``: the name of the ColumnDataSource that contains the columns you
-  just referenced for the ``x`` and ``y`` arguments.
+To modify the data of an existing ColumnDataSource, update the ``.data``
+property of your ColumnDataSource object:
+
+* To add a new column to an existing ColumnDataSource:
+
+  .. code-block:: python
+
+    new_sequence = [8, 1, 4, 7, 3]
+    source.data["new_column"] = new_sequence
+
+  The length of the column you are adding must match the length of the existing
+  columns.
+
+* To replace data in an existing column within your ColumnDataSource, assign
+  your new value to the respective index position:
+
+  .. code-block:: python
+
+    source.data["new_column"][2] = 5
+
+* To replace all data in an existing ColumnDataSource, assign the ``.data``
+  property an entirely new dict:
+
+  .. code-block:: python
+
+    source.data = new_dict
 
 .. note::
     Bokeh assumes that all columns in a ColumnDataSource each have the same
-    length at all times. For this reason, make sure to always update all columns
-    of a ColumnDataSource at the same time.
+    length at all times. For this reason, there are some things to keep in mind:
+
+    * When creating a new ColumnDataSource: Make sure all sequences of values
+      you pass into a ColumnDataSource are the same length.
+    * When adding a new column to an existing ColumnDataSource by updating the
+      ``.data`` property: Make sure your additional column is the same length as
+      the columns already in the ColumnDataSource.
+    * When updating data in a way that changes the length of any column in a
+      ColumnDataSource: You must update all columns at once by passing an
+      entirely new dict containing columns with the same length. It is not
+      possible to update column lengths one column at a time.
```

Comment:
With this new section I thikn we an just fold these individual points into the appropriate place above, and not have a note at all?

Reply:
Since this seems to be a problem for so many users, I think it makes sense to really highlight this information. In my next commit, I have added a note to the appropriate places in "Creating a ColumnDataSource" and "Modifying a ColumnDataSource".